crates/driver/src/domain/competition/auction.rs

```diff
@@ -76,12 +62,95 @@ impl Auction {
+            // are the most niche and rarely used.
+            let class = match order.kind {
+                competition::order::Kind::Market => 2,
+                competition::order::Kind::Limit { .. } => 1,
+                competition::order::Kind::Liquidity => 0,
+            };
+            std::cmp::Reverse((
+                class,
+                // TODO No need to expose Tokens anymore, just use the auction
+                // If the orders are of the same kind, then sort by likelihood of fulfillment
+                // based on token prices.
+                order.likelihood(&self.tokens),
+            ))
+        });
+
+        // Fetch balances of each token for each trader.
+        let tokens_by_trader = self
+            .orders
+            .iter()
+            .flat_map(|order| {
+                [
+                    (order.trader(), order.sell.token),
+                    (order.trader(), order.buy.token),
+                ]
+            })
+            .unique()
+            .collect_vec();
+        let mut balances: HashMap<(order::Trader, eth::TokenAddress), eth::TokenAmount> =
+            join_all(tokens_by_trader.into_iter().map(|(trader, token)| {
+                async move {
+                    let balance = eth
+                        .balance_of(trader.into(), token)
+                        .await
+                        // If fetching the balance of this trader fails, don't filter out their
+                        // orders. Pretend like they have infinite balance.
+                        .unwrap_or(eth::U256::MAX.into());
+                    ((trader, token), balance)
+                }
+            }))
+            .await
+            .into_iter()
+            .collect();
+
+        // Filter out orders which the trader doesn't have enough balance to pay for.
+        self.orders.retain(|order| {
+            // Partial orders don't require the trader to hold the full balance, since they
+            // don't have to be fully fulfilled.
+            if order.is_partial() {
+                return true;
+            }
+
+            // Sell amounts are withdrawn from the trader's balance when the order settles.
+            // In case the trader doesn't have enough balance to pay for the order, filter
+            // it out.
+            let sell_balance = balances
+                .get_mut(&(order.trader(), order.sell.token))
+                .unwrap();
+            match sell_balance.0.checked_sub(order.sell.amount.into()) {
+                Some(remaining) => sell_balance.0 = remaining,
+                None => return false,
+            }
+
+            // Buy amounts are deposited into the trader's balance when the order settles.
+            let buy_balance = balances
+                .get_mut(&(order.trader(), order.buy.token))
+                .unwrap();
+            buy_balance.0 = buy_balance.0.saturating_add(order.buy.amount.into());
```

Comment:
This doesn't work because all transfers happen at the start of the settlement before any additional buy amount shows up.